espnet2/samplers/build_batch_sampler.py

```diff
@@ -0,0 +1,94 @@
+from typing import List
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+
+from typeguard import check_argument_types
+from typeguard import check_return_type
+
+from espnet2.samplers.abs_sampler import AbsSampler
+from espnet2.samplers.batch_bin_sampler import BatchBinSampler
+from espnet2.samplers.constant_batch_sampler import ConstantBatchSampler
+from espnet2.samplers.constant_sorted_batch_sampler import ConstantSortedBatchSampler
+from espnet2.samplers.sequence_batch_sampler import SequenceBatchSampler
+
+
+def build_batch_sampler(
+    type: str,
+    batch_size: int,
+    batch_bins: int,
+    shape_files: Union[Tuple[str, ...], List[str]],
+    fold_lengths: Sequence[int] = (),
+    sort_in_batch: str = "descending",
+    sort_batch: str = "ascending",
+    drop_last: bool = False,
+    min_batch_size: int = 1,
+    padding: bool = True,
+) -> AbsSampler:
+    """Helper function to instantiate BatchSampler
+
+    Args:
+        type: mini-batch type. "constant", "seq", "bin", or, "frame"
```

Comment:
Here are readability issues and the discrepancy between this `type` names and the implementation class names.  And we can forget espnet1 compatibility and reconsider better names. For example

- `const_no_sort` `ContantBatchSampler` => `unsorted` `UnsortedBatchSampler`
- `const` `ConstantSortedBatchSampler` => `sorted` `SortedBatchSampler`
- `seq` `SequenceBatchSampler` => `target-length` `TargetLengthBatchSampler`
- `frame` NA => `source-length` `SourceLengthBatchSampler`
- `bins` `BatchBinSampler` => `source-target-numel` `SourceTargetNumelBatchSampler`

ref: [numel](https://pytorch.org/docs/stable/torch.html#torch.numel)

Reply:
I didn't care about the naming. `unsorted` and `sorted` sound good.

What does `target-length` means? SequenceBatchSampler takes into account all input sequences i.e.  the input and the target?

Reply:
> SequenceBatchSampler takes into account all input sequences i.e. the input and the target?

Oh, yes. You are right. The `seq` option is too complicated!

Reply:
hmm, I don't have good name.
seq and frame have similar behaviour and we need to name explicitly...

Or, do we rather to discard seq and frame mode?
"bins" and "constant" may be enough? Actually. I'm sicking of deciding `max_length` values. They looks  ugly in new recipes.


https://github.com/espnet/espnet/blob/4799542be4a0579752f953627d1cf8cd3b13668c/egs2/TEMPLATE/asr1/asr.sh#L763-L781

Reply:
@ShigekiKarita @sw005320 Sounds odd?

- `const_no_sort` `ContantBatchSampler` => `unsorted` `UnsortedBatchSampler`
- `const` `ConstantSortedBatchSampler` => `sorted` `SortedBatchSampler`
- `seq` `SequenceBatchSampler` => `folded` `FoldedLengthBatchSampler` 
- `frame` `NA` => `len` `LengthBatchSampler`
-  `bins` `BatchBinSampler` => `numel` `NumElementsBatchSampler`

Reply:
All the method folds the original batchsize anyway.
How about distinguishing it with the **unit** to use folding?

Reply:
@sw005320 Sorry, I'm not sure what you mean. Could you say more?

Reply:
all methods first define the unit (e.g., input/output length, number of elements of input and output), and use this unit as a threshold to define the batch size.
Probably I'm wrong.

Reply:
seq mode is not matched this pattern?

```
If any feats lengths in the mini-batch are more than N * fold_length:
    => bs // N
```

Reply:
OK.
Then you can keep this name.